### ISSUE TITLE
How to use fp16

### ISSUE THREAD

**Issue by JBoRu** [HKUNLP/UnifiedSKG#31]
Hi,
   I find the `deepspeed` directory in the project. And I want to know how to train with `fp16`? Now, I  just add the `--fp16`  in the command, but it seems not to work.

**Comment by Timothyxxx** [comment]
Hi, 

Could you provide the command you used, and what do you mean by not to work(error, or nan in loss...)?

Thanks!

**Comment by JBoRu** [comment]
Hi,
 My command is like this, it just adds the extra `--fp16 `, but the GPU memory seems not to reduce. 

> CUDA_VISIBLE_DEVICES=5,6,7 /home/jiangjinhao/anaconda3/envs/USKG-pt1.8/bin/python3 -W ignore -m torch.distributed.launch --nproc_per_node 3 --master_port 1237 train.py --seed 2 --cfg Salesforce/T5_large_finetune_webqsp_retrieval.cfg --run_name T5_large_finetune_webqsp_retrieval_lower_dfs --logging_strategy steps --logging_first_step true --logging_steps 40 --evaluation_strategy steps --eval_steps 40 --metric_for_best_model eval_META_TUNING/webqsp_retrieval.cfg/Hits@1 --greater_is_better true --save_strategy steps --save_steps 40 --save_total_limit 1 --load_best_model_at_end --gradient_accumulation_steps 30 --num_train_epochs 15 --adafactor true --learning_rate 5e-5 --do_train --do_eval --do_predict --predict_with_generate --output_dir output/T5_large_finetune_webqsp_retrieval_lower_dfs --overwrite_output_dir --per_device_train_batch_size 1 --per_device_eval_batch_size 1 --generation_num_beams 4 --generation_max_length 28 --input_max_length 1024 --ddp_find_unused_parameters False --fp16

At the same time, if I set the `--ddp_find_unused_parameters True` , the log will output `there are no unused parameters`. So, what role of this arg.

**Comment by Timothyxxx** [comment]
Hi!

1, Oh I remember we used similar command before, and also have little effect on speed(We used 8*Tesla V100 32GB cluster at that time). And we didn't spend time exploring why that happen. Maybe you can try deepspeed [as we showed](https://github.com/HKUNLP/UnifiedSKG#training).
2, `--ddp_find_unused_parameters` is a arg for PyTorch to decide whether to find unused parameters, it will make no difference in result, only cause slight speed difference.

**Comment by Timothyxxx** [comment]
btw, your loss got decreased normally when using `fp16` on T5? which version of `transformers` did you used?

**Comment by JBoRu** [comment]
Hi, 
   Sorry that I don't check the loss for the command with `fp16` because I still meet the `CUDA OUT OF MEMORY` error. And the version of `transformers` is 4.9.2, and all of the others can work fine, except that I can't prefix tuning T5-3B on RTX 3090(24G) with batch_size=1.

**Comment by Timothyxxx** [comment]
24GB GPU may not enough for the T5-3B training, only enough for inference(bsz could be set up to 2 when 24 GB GPU memory). I recommend give deepspeed a try but I am not sure again 24 GB GPU will be enough for 3B.

**Comment by JBoRu** [comment]
24 GB is still not enough for prefixtuning T5-3B. I will try a more large memory GPU.
